draco/analysis/radiometer_test.py

```diff
@@ -0,0 +1,238 @@
+        mask.weights[:] = var_mask
+
+        return weights, mask
+    
+    def stack_auto_noise(self, data, inputmap):
+
+        data.redistribute('freq')
+        nfreq = data.vis.local_shape[0]
+
+	#Relative cylinder weighting
+	N_XX   = 0
+	N_YY   = 0
+        #Intercylinder baselines
+	Nbl_XX = 0
+	Nbl_YY = 0	
+
+        #Shape of data: (nfreq, nprod, ntime)
+
+        #Getting indices of autocorrelations 
+        #separately for the two polarizations
+        auto_stack_id = [] #IDs for autocorrelations 
+        ind_XX = [] #Index of polarization within the selection
+        ind_YY = []
+        auto_pol_pos = 0 #Counter for getting the index
+        for ii in range(len(data.index_map["prod"][:])):
+            feedi, feedj = self.telescope.uniquepairs[ii]
+            pol_0        = self.telescope.polarisation[feedi]
+            pol_1        = self.telescope.polarisation[feedj]
+            if feedi==feedj:
+                auto_stack_id.append(ii)
+                if pol_0=='X' and pol_1=='X':
+                    ind_XX.append(auto_pol_pos)
+                elif pol_0=='Y' and pol_1=='Y':
+                    ind_YY.append(auto_pol_pos)
+                auto_pol_pos +=1
+
+        bflag = (data.weight.data[:,auto_stack_id,:] == 0.0)
+        bvis  = data.vis.data[:,auto_stack_id:,]
+	bfrac = data.flags['frac_lost'].data
+	binp  = data.flags['inputs']
+        time_auto = data.time
+
+	#Derive weights for each cylinder, based on the number of active feeds
+	live_feed_cyl_XX = np.zeros((ncyl, len(time_auto)))
+	live_feed_cyl_YY = np.zeros((ncyl, len(time_auto)))
+
+	for i in range(ncyl):
+	    live_feed_cyl_XX[i] = live_feeds(chr(i+ord('A')),'XX',binp)
+	    live_feed_cyl_YY[i] = live_feeds(chr(i+ord('A')),'YY',binp)
+
+	#Average autos across cylinders
+	vis_avg_XX = np.zeros((nfreq,len(time_auto)), dtype='complex64')
+	vis_avg_YY = np.zeros((nfreq,len(time_auto)), dtype='complex64')
+
+	#Averaging for pol XX (multiplying visibilities pairwise between different cylinders)
+	for i in ind_XX:
+	    for j in ind_XX:
+		if i >= j: #Avoid double counting
+		    continue 
+		
+                n_feeds_cyl_XX = live_feeds(chr(i+ord('A')),'XX',binp)*live_feeds(chr(j+ord('A')),'XX',binp) #Total number of feeds in this product
+		vis_avg_XX +=  bvis[:,i,:]*bvis[:,j,:] * (n_feeds_cyl_XX/feed_qnt**2) # Product of visibilities weighted by fraction of non-flagged feeds
+		
+                Nbl_XX += n_feeds_cyl_XX #Counter for total number of feeds averaged
+                N_XX += n_feeds_cyl_XX/feed_qnt**2 #Effective number of total averages, max: 6 
+		
+	#Averaging for pol YY
+	for i in ind_YY:
+	    for j in ind_YY:
+		if i >= j: 
+		    continue 
+		
+                n_feeds_cyl_YY = live_feeds(chr(i+ord('A')-len(ind_XX)),'YY',binp)*live_feeds(chr(j+ord('A')-len(ind_XX)),'YY',binp)
+		vis_avg_YY +=  bvis[:,i,:] * bvis[:,j,:] * n_feeds_cyl_YY/feed_qnt**2
+		
+                Nbl_YY += n_feeds_cyl_YY
+		N_YY += n_feeds_cyl_YY/feed_qnt**2
+
```

Comment:
I think lines 185 - 220 could be replaced with something like this:

```
radiometer = np.zeros((nfreq, npol, ntime), dtype=np.float32)
radiometer_counter = np.zeros((nfreq, npol, ntime), dtype=np.float32)

for ii, (ai, pi) in enumerate(zip(auto_input, auto_pol)):
    
    for jj, (aj, pj) in enumerate(zip(auto_input, auto_pol)):
        
        if pi == pj and ii > jj:
            # Avoid double counting
            continue
            
        if not self.include_intracyl and (self.telescope.feedpositions[ai, 0] == self.telescope.feedpositions[aj, 0]):
            # Exclude intracylinder
            continue
            
        if pi != pj and (ai > aj):
            # Deals with oddities in cross-pol intracylinder baselines
            if self.include_intracyl and (self.telescope.feedpositions[ai, 0] == self.telescope.feedpositions[aj, 0]):
                pass
            else:
                continue
            
        pp = upol.index(auto_pol[ii] + auto_pol[jj])
        
        n2 = num_live_feed[np.newaxis, ii, :] * bflag[:, ii, :] * num_live_feed[np.newaxis, jj, :] * bflag[:, jj, :]
        
        radiometer_counter[:, pp, :] += n2
        radiometer[:, pp, :] += n2 * bvis[:, ii, :] * bvis[:, jj, :]```

Reply:
so that you don't have to repeat the code and also so that we can handle both co- and cross-polar baselines.